### DISCUSSION TITLE
PointPerfect MQTT Correction

### DISCUSSION THREAD

**Discussion by drodriguez-dlc** [semuconsulting/PyGPSClient#93] [General]
I'm using example/client.py with pygnsutils GNSSSocketServer and trying to implement GNSSMQTTClient. I start the server with this command
`gnssserver --inport /dev/ttyGPS --baudrate 115200 --hostip 0.0.0.0 --outport 50011`

Then start the MQTT connection
`gnssmqttclient -C c3938143-fe52-xxxx-b2da-xxxxxxxxxxxx -S pp.services.u-blox.com -P 8883 -R us --output /dev/ttyGPS`
I can see my device connected on PointPerfect(PP), but I'm unsure of how to actually send the correction data to my ZED-F9P. 

Utilizing pygpsclient UI, I can connect to my server at 0.0.0.0:50011 and then connect to my MQTT broker to retrieve the correction data and I can eventually get an RTK fix. 

How can I replicate this pygpsclient functionality using the example client, gnssserver, and gnssmqttclient? 
Would GNSSStreamer be better suited for this task?

**Comment by semuadmin** [comment]
Hi @drodriguez-dlc 

Unfortunately you can't use gnssmqttclient to output directly to a serial device (e.g. `/dev/ttyGPS`) - you need to open it as a writeable stream first. The gnssmqttclient `--output` parameter accepts any of the following writeable output classes:

- serial stream (Serial)
- binary file stream (BufferedWriter)
- text file stream (TextIOWrapper)
- queue (Queue)
- socket (Socket)

The way this works 'under the covers' in PyGPSClient is that it writes the output from gnssmqttclient to a queue, and a separate threaded process reads this queue and sends the data to the receiver.

The [rtk_example.py](https://github.com/semuconsulting/pygnssutils/blob/main/examples/rtk_example.py) does something similar but for NTRIP clients rather than SPARTN, but you could adapt it quite easily to use `GNSSMQTTClient` rather than `GNSSNTRIPClient`. If you need any guidance doing this, get back to me.

**Reply by drodriguez-dlc** [reply]
Very much appreciate the quick reply! I'll give this a shot this evening.

**Reply by drodriguez-dlc** [reply]
I was able to modify [rtk_exmaple.py](https://github.com/semuconsulting/pygnssutils/blob/main/examples/rtk_example.py) to output its data to the send_queue. Thanks again for the help!
I do have one more question that is not related to the original question. Do you prefer a new thread for that?

**Reply by semuadmin** [reply]
If it's an unrelated subject, create a new thread. Thanks. Could I ask you to mark this thread as 'Answered' if you're happy with the resolution.
